### ISSUE TITLE
Symbols loop

### ISSUE THREAD

**Issue by diorcety** [swig/swig#366]
I'm currently trying to wrap a subset of eigen. Swig starts to hang at some points here the debug logs

http://dl.free.fr/gHqc3wL0K

**Comment by ojwb** [comment]
That link doesn't seem to let me see the file.

Anyway, what we really need is a reduced self-contained example demonstrating the issue.

**Comment by diorcety** [comment]
```
%module(package="eigen3", directors="1") eigen3

#include <Eigen/Core>

#include <Eigen/src/Core/util/DisableStupidWarnings.h>

#define __GNUC__
%include <Eigen/src/Core/util/Macros.h>
#undef __GNUC__
#undef EIGEN_ALIGN_TO_BOUNDARY
#define EIGEN_ALIGN_TO_BOUNDARY(n) __attribute__((aligned(n)))

%include <Eigen/src/Core/util/Constants.h>
%include <Eigen/src/Core/util/ForwardDeclarations.h>

%include <Eigen/src/Core/MatrixBase.h>
```

/usr/local/bin/swig -I/usr/include/eigen3   -c++ -python -interface libeigen3_python -o eigen3_wrap.cpp test.i

**Comment by vadz** [comment]
This is hardly self-contained... Any chance of putting just the necessary [to make it hang] parts of Eigen inline here?

**Comment by diorcety** [comment]
Ok i will try to minimize the test again ... not so easy.
Btw it's really difficult to use existing full "templatized" library like eigen

**Comment by vadz** [comment]
Yes, of course, it's not easy. But, because you at least know it, you have a much better chance of doing it than, for example, I do...

Anyhow, good luck and thanks in advance!

**Comment by ojwb** [comment]
@diorcety Is this still an issue with current SWIG?

If so, could you try to come up with a reduced testcase?

**Comment by ojwb** [comment]
Installing `libeigen3-dev` on debian unstable for the headers, it does still seem to happen.  It appears to be endlessly recursing via what's currently line 992 in `Source/Swig/symbol.c`:

```
      n = _symbol_lookup(dname, symtab, check);
```

I've never worked out how to usefully inspect DOH objects from gdb though so not sure exactly what up without adding `Printf` calls.

I managed to hack the testcase down to just:

```
%module x
template<int R> class M;
typedef M<2 >= 1 ? 2 : 1> S;
```

Weirdly, the `>=` seems to really matter - changing it to `!=` or even `<=` avoids the infinite recursion.  It's not the result of the comparison mattering as `1 >= 2` also fails while `1 <= 2` works.

Probably not going to dig deeper for now so if anyone else wants to please do.

**Comment by ojwb** [comment]
My reduced testcase processes with git master but still loops with 4.1.0.  Using `git bisect` points to this being fixed by ca5c68e5441fa8334381bd26ea2ff5132c26e6ab (which was a fix for #1037).  That means this is fixed, and the fix will be in SWIG 4.2.0.

**Comment by wsfulton** [comment]
@ojwb, the DOH objects are easily viewable in gdb if you use swig.gdb, see https://htmlpreview.github.io/?https://github.com/swig/swig/blob/master/Doc/Devel/internals.html#7. SWIG is near impossible to debug without this.
